process/block/metablock.go

```diff
@@ -1618,6 +1618,42 @@ func (mp *metaProcessor) createEpochStartForMetablock(metaBlock *block.MetaBlock
 	return epochStart, nil
 }
 
+func (mp *metaProcessor) getLastFinalizedMetaHashForShard(shardHdr *block.Header) ([]byte, []byte, error) {
+	var lastMetaHash []byte
+	var lastFinalizedMetaHash []byte
+
+	numAddedMetas := len(shardHdr.MetaBlockHashes)
+	if numAddedMetas > 0 {
+		lastMetaHash = shardHdr.MetaBlockHashes[numAddedMetas-1]
+	}
+
+	if numAddedMetas > 1 {
+		lastFinalizedMetaHash = shardHdr.MetaBlockHashes[numAddedMetas-1]
+		return lastMetaHash, lastFinalizedMetaHash, nil
+	}
+
+	for currentHdr := shardHdr; currentHdr.GetNonce() > 0 && currentHdr.GetEpoch() == shardHdr.GetEpoch(); {
+		oldShardHdr, err := process.GetShardHeader(currentHdr.GetPrevHash(), mp.dataPool.ShardHeaders(), mp.marshalizer, mp.store)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if len(oldShardHdr.MetaBlockHashes) == 0 {
+			currentHdr = oldShardHdr
+			continue
+		}
+
+		lastFinalizedMetaHash = oldShardHdr.MetaBlockHashes[len(oldShardHdr.MetaBlockHashes)-1]
+		if len(lastMetaHash) == 0 {
+			lastMetaHash = lastFinalizedMetaHash
```

Comment:
I think you shoud add here:
if len(oldShardHdr.MetaBlockHashes) <= 1 {
continue
}

lastFinalizedMetaHash = oldShardHdr.MetaBlockHashes[len(oldShardHdr.MetaBlockHashes)-2]